htmresearch/algorithms/sparse_net.py

```diff
@@ -0,0 +1,352 @@
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+Implementation of Bruno Olshausen's sparse coding algorithm.
+It relies on the formulation developed by Olshausen and Field (1996), slightly
+modified so that it uses a Locally Competitive Algorithm (LCA) to compute the
+coefficients, rather than a vanilla gradient descent, as proposed by Rozell
+et al. (2008).
+
+Images have to be flattened, else the specific ImageSparseNet has to be used.
+
+Example use:
+  # create and train network
+  net = SparseNet(inputDim=64,
+                  outputDim=64,
+                  verbosity=2,
+                  numIterations=1000,
+                  batchSize=100)
+  data = np.random.randn(64, 700)
+  net.train(data)
+
+  # visualize loss history and basis
+  net.plotLoss(filename="loss_history.png")
+  net.plotBasis(filename="basis_functions.png")
+
+  # encode data (works with single point or dataset)
+  encodings = net.encode(dataToEncode)
+
+"""
+
+import random
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+class SparseNet(object):
+  """
+  Default SparseNet implementation, which works on most common data types.
+
+  It provides public methods for training and encoding data, as well as methods
+  for plotting the network's basis and loss history.
+  """
+
+  def __init__(self,
+               inputDim=64,
+               outputDim=64,
+               batchSize=100,
+               numIterations=1000,
+               numLcaIterations=75,
+               learningRate=2.0,
+               decayCycle=100,
+               learningRateDecay=1.0,
+               lcaLearningRate=0.1,
+               thresholdDecay=0.95,
+               minThreshold=0.1,
+               thresholdType='soft',
+               verbosity=0,
+               showEvery=500,
+               seed=42):
+    """
+    Initializes the SparseNet.
+    :param inputDim:                (int)   (Flattened) dimension of input data
```

Comment:
OK. Then I guess the _getDataBatch method in the base class should be abstract? And be implemented in sub-classes specific to data types?